core/templates/pages/teach-page/teach-page.component.spec.ts

```diff
@@ -15,112 +15,197 @@
-
-    ctrl.onTabClick('participation');
+  it('should click on participation tab', (() => {
+    component.ngOnInit();
+    expect(component.activeTabName).toBe('teach');
 
-    // Function setTimeout is being used here in order to wait onhashchange
-    // event to finish. setTimeout is executed only after call stack is empty.
-    // Ref: https://developer.mozilla.org/en-US/docs/Web/JavaScript/EventLoop
-    setTimeout(function() {
-      expect(windowRef.nativeWindow.location.hash).toBe('#participation');
-      expect(ctrl.activeTabName).toBe('participation');
-      done();
-    });
-  });
+    component.onTabClick('participation');
+    expect(windowRef.nativeWindow.location.hash).toBe('#participation');
+    expect(component.activeTabName).toBe('participation');
+  }));
 
-  it('should activate teach tab on init', function() {
+  it('should activate teach tab on init', () => {
     windowRef.nativeWindow.location.hash = '#teach';
 
-    ctrl.$onInit();
+    component.ngOnInit();
 
     expect(windowRef.nativeWindow.location.hash).toBe('#teach');
-    expect(ctrl.activeTabName).toBe('teach');
+    expect(component.activeTabName).toBe('teach');
   });
 
-  it('should activate participation tab on init', function() {
+  it('should activate participation tab on init', () => {
     windowRef.nativeWindow.location.hash = '#participation';
 
-    ctrl.$onInit();
+    component.ngOnInit();
 
     expect(windowRef.nativeWindow.location.hash).toBe('#participation');
-    expect(ctrl.activeTabName).toBe('participation');
+    expect(component.activeTabName).toBe('participation');
   });
 
-  it('should get static image url', function() {
-    expect(ctrl.getStaticImageUrl('/path/to/image')).toBe(
+  it('should get static image url', () => {
+    expect(component.getStaticImageUrl('/path/to/image')).toBe(
       '/assets/images/path/to/image');
   });
 
-  it('should apply to teach with oppia', function() {
-    var applyToTeachWithOppiaEventSpy = spyOn(
-      SiteAnalyticsService, 'registerApplyToTeachWithOppiaEvent')
+  it('should apply to teach with oppia', () => {
+    const applyToTeachWithOppiaEventSpy = spyOn(
+      siteAnalyticsService, 'registerApplyToTeachWithOppiaEvent')
       .and.callThrough();
 
-    ctrl.$onInit();
+    component.ngOnInit();
     spyOnProperty(windowRef, 'nativeWindow').and.returnValue({
-      location: ''
+      location: {
+        href: ''
+      }
     });
-    ctrl.onApplyToTeachWithOppia('/login');
-    $timeout.flush(150);
-
-    expect(windowRef.nativeWindow.location).toBe(
-      'https://goo.gl/forms/0p3Axuw5tLjTfiri1');
-    expect(applyToTeachWithOppiaEventSpy).toHaveBeenCalled();
+    component.onApplyToTeachWithOppia();
+    setTimeout(() => {
+      fixture.detectChanges();
+      expect(windowRef.nativeWindow.location.href).toBe(
+        'https://goo.gl/forms/0p3Axuw5tLjTfiri1');
+      expect(applyToTeachWithOppiaEventSpy).toHaveBeenCalled();
+    }, 500);
   });
```

Comment:
```suggestion
    tick(150);
    fixture.detectChanges();
    expect(windowRef.nativeWindow.location.href).toBe(
      'https://goo.gl/forms/0p3Axuw5tLjTfiri1');
    expect(applyToTeachWithOppiaEventSpy).toHaveBeenCalled();
  }));
```

Reply:
Done

Reply:
Why setTimeout didn't work and tick worked (with less timeout)?

Reply:
@srijanreddy98 said that the function (in the component file) shouldn't have been using setTimeout in the first place. The tick thing is just to simulate that time passing. The reason setTimeout didn't work here (I think, I'm not sure) is because the setTimeout in the spec function was getting executed before the one in the component function. This explains why we have more success with using higher timeouts, but that too is not a reliable fix.

Reply:
Also, using setTimeout ended the test cycle without running other tests, contrary to usual behaviour wherein the other tests would continue running. I'm actually not sure why this happened

Reply:
> Why setTimeout didn't work and tick worked (with less timeout)?

Not just less time, it is the exact time specified in component.ts file. setTimeout is added to the callback queue after the specified time and that makes its behavior unpredictable when trying to "match" another setTimeout. `Tick(x)` can be considered as an extension of `process.nextTick()`. Hence it is more reliable.